### PR TITLE
Add libsawtooth and stores

### PR DESCRIPTION
**Pull request by ltseeley**
- Add the libsawtooth crate that will provide various components used by the Sawtooth validator.
- Add the `OrderedStore` trait, which defines an interface for a key/value store that is indexed by a type with total ordering.
- Add a BTreeMap-backed implementation of the `OrderedStore` trait for an in-memory ordered store.
- Add the `TransactionReceiptStore`, which is a wrapper around an `OrderedStore` that can be used to store transaction receipts, keyed by transaction IDs, in order.
- Add the `RedisOrderedStore` struct, which is an implementation of the `OrderedStore` trait that uses a redis database.
- Add the `LmdbOrderedStore`, which is an implementation of the `OrderedStore` trait that uses an LMDB database.

### CODE REVIEW

File: libsawtooth/src/store/lmdb.rs [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+            OrderedStoreError::InitializationFailed(format!("invalid filepath: {:?}", filepath))
+        })?;
+
+        let mut builder = lmdb::EnvBuilder::new().map_err(|err| {
+            OrderedStoreError::InitializationFailed(format!(
+                "failed to initialize environment builder: {}",
+                err
+            ))
+        })?;
+        builder.set_maxdbs(3).map_err(|err| {
```
Comment (vaporos) [inline libsawtooth/src/store/lmdb.rs]: Maybe const this value.

Comment (ltseeley) [reply libsawtooth/src/store/lmdb.rs]: Done

**Comment by ltseeley** [comment]
Put modules and sub-modules behind experimental features

File: libsawtooth/src/store/receipt_store/mod.rs [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+use transact::protocol::receipt::TransactionReceipt;
+
+use crate::store::OrderedStore;
+
+pub use self::error::TransactionReceiptStoreError;
+
+/// `TransactionReceiptStore` is a wrapper around an `OrderedStore` that stores
+/// `TransactionReceipt`s keyed by transaction IDs (`String`s) and indexed by `usize`s to provide
+/// ordering of transactions.
+pub struct TransactionReceiptStore(Box<dyn OrderedStore<String, TransactionReceipt, usize> + Send>);
```
Comment (peterschwarz) [inline libsawtooth/src/store/receipt_store/mod.rs]: This should be `u64`, if it is the equivalent of block height

**Pull request merged**